zio/ndjsonio/typeparser.go

```diff
@@ -0,0 +1,322 @@
+package ndjsonio
+
+import (
+	"bytes"
+	"errors"
+	"fmt"
+	"strconv"
+	"strings"
+
+	"github.com/brimsec/zq/pkg/nano"
+	"github.com/brimsec/zq/zcode"
+	"github.com/brimsec/zq/zio/zeekio"
+	"github.com/brimsec/zq/zng"
+	"github.com/brimsec/zq/zng/resolver"
+	"github.com/buger/jsonparser"
+)
+
+type typeStats struct {
+	BadFormat            int
+	FirstBadLine         int
+	DescriptorNotFound   int
+	IncompleteDescriptor int
+	MissingPath          int
+}
+
+type typeParser struct {
+	lineNo        int
+	zctx          *resolver.Context
+	tr            typeRules
+	defaultPath   string
+	stats         *typeStats
+	typeInfoCache map[int]*typeInfo
+}
+
+var (
+	ErrDescriptorNotFound = errors.New("descriptor not found")
+	ErrMissingPath        = errors.New("missing path field")
+	ErrBadFormat          = errors.New("bad format")
+)
+
+// Information about the correspondence between the flattened structure
+// of a JSON object and its zng representation (which may include
+// nested record fields). The two descriptors here represent the same data
+// in the same order, flatDescriptor describes the data as it appears in
+// JSON, descriptor describes it as it appears in zng values.
+type typeInfo struct {
+	descriptor *zng.TypeRecord
+	flatDesc   *zng.TypeRecord
+	path       []byte
+	jsonVals   []jsonVal
+}
+type jsonVal struct {
+	val []byte
+	typ jsonparser.ValueType
+}
+
+func getUnsafeDefault(data []byte, defaultValue string, key string) (string, error) {
+	val, err := jsonparser.GetUnsafeString(data, key)
+	if err != nil {
+		// This is always a KeyPathNotFoundError, including if the json was invalid.
+		if defaultValue == "" {
+			return "", jsonparser.KeyPathNotFoundError
+		}
+		return defaultValue, nil
+	}
+	return val, nil
+}
+
+func lookupTypeInfo(zctx *resolver.Context, desc *zng.TypeRecord, path string) *typeInfo {
```

Comment:
nit: this is doing more than a lookup, maybe call it makeTypeInfo or newTypeInfo?

Reply:
👍